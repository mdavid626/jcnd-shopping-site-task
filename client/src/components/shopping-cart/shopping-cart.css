.ShoppingCart {
  background-color: #f2f2f2;
  border-radius: 5px;
  padding: var(--space3);
  display: flex;
  flex-direction: column;
  gap: var(--space3);
  min-width: 700px;
  min-height: 400px;
}

.ShoppingCart-header {
  font-size: 22px;
}

.ShoppingCart-content {
  display: flex;
  flex-direction: column;
  flex: 1;
}

.ShoppingCart-close {
  cursor: pointer;
  color: #b3005e;
}

.ShoppingCart-items {
  display: flex;
  flex-direction: column;
  gap: var(--space3);
  overflow-y: scroll;
  max-height: 280px;
  padding: 0 var(--space2) 0 0;
  flex: 1;
}

.ShoppingCart-total {
  align-self: flex-end;
  font-size: 18px;
}

.ShoppingCart-actions {
  display: flex;
  gap: var(--space3);
  align-items: center;
  align-self: flex-end;
}

.ShoppingCart-placeOrder {
  cursor: pointer;
  font-size: 18px;
  background-color: var(--primary-color);
  padding: var(--space2) var(--space3);
  color: #f2f2f2;
  border-radius: 3px;
}
