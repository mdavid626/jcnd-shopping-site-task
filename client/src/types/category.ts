export type Category = 'vegetables' | 'fruits' | 'cheese';
