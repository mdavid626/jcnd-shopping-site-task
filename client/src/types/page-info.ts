export type PageInfo = {
  page: number;
  numberOfPages: number;
};
