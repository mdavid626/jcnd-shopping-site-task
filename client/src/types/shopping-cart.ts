export type ShoppingCartItem = {
  productId: string;
  amount: number;
};
