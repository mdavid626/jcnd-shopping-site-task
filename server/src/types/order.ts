import { Types } from 'mongoose';

export type OrderItem = {
  productId: Types.ObjectId;
  amount: number;
};

export type Order = {
  items: OrderItem[];
  createdAt: Date;
  updatedAt: Date;
};

export type OrderRequestItem = {
  productId: string;
  amount: number;
};
