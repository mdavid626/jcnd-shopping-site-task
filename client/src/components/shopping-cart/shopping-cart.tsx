import React from 'react';
import {
  usePlaceOrder,
  useShoppingCart,
} from '../../hooks/shopping-car-thooks/shopping-cart-hooks';
import { ShoppingCartItem } from '../../types/shopping-cart';
import Currency from '../currency/currency';
import ShoppingCartListItem from '../shopping-cart-list-item/shopping-cart-list-item';
import './shopping-cart.css';

const calculateTotal = (shoppingCart: ShoppingCartItem[]) =>
  shoppingCart.reduce(
    (acc, item) => acc + item.product.priceInCents * item.amount,
    0
  );

const ShoppingCart: React.FC<{ onClose: () => void }> = ({ onClose }) => {
  const [shoppingCart] = useShoppingCart();
  const [placeOrder, isPlacingOrder] = usePlaceOrder();
  const handlePlaceOrderClick = async () => {
    await placeOrder();
    onClose();
  };
  return (
    <div className="ShoppingCart">
      <div className="ShoppingCart-header">Shopping Cart</div>
      <div className="ShoppingCart-content">
        {shoppingCart.length > 0 ? (
          <div className="ShoppingCart-items">
            {shoppingCart.map((item) => (
              <ShoppingCartListItem
                key={item.product.id}
                shoppingCartItem={item}
              />
            ))}
          </div>
        ) : (
          <div>Your shopping cart is empty</div>
        )}
      </div>
      {shoppingCart.length > 0 && (
        <div className="ShoppingCart-total" data-testid="ShoppingCart-total">
          Total: <Currency priceInCents={calculateTotal(shoppingCart)} />
        </div>
      )}
      <div className="ShoppingCart-actions">
        <div onClick={() => onClose()} className="ShoppingCart-close">
          close
        </div>
        {shoppingCart.length > 0 && !isPlacingOrder && (
          <div
            onClick={handlePlaceOrderClick}
            className="ShoppingCart-placeOrder"
          >
            Place order
          </div>
        )}
        {isPlacingOrder && (
          <div className="ShoppingCart-placeOrder">Placing order...</div>
        )}
      </div>
    </div>
  );
};

export default ShoppingCart;
