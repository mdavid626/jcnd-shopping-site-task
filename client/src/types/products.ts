import { PageInfo } from './page-info';

export type ProductCategory = 'VEGETABLES' | 'FRUITS' | 'CHEESE';

export type Product = {
  id: string;
  price: number;
  name: string;
  inStock: number;
};

export type ProductsQueryResult = {
  products: { nodes: Product[]; pageInfo: PageInfo };
};
