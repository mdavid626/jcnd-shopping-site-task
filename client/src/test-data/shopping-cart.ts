import { ShoppingCartItem } from '../types/shopping-cart';
import { testProduct1, testProduct2 } from './products';

export const testShoppingCartItem1: ShoppingCartItem = {
  product: testProduct1,
  amount: 3,
};

export const testShoppingCartItem2: ShoppingCartItem = {
  product: testProduct2,
  amount: 36,
};

export const testShoppingCart1: ShoppingCartItem[] = [
  testShoppingCartItem1,
  testShoppingCartItem2,
];
