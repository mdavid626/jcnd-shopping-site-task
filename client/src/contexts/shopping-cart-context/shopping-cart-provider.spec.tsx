import { act, render, RenderOptions } from '@testing-library/react';
import React, { ReactElement, useContext } from 'react';
import {
  testShoppingCart1,
  testShoppingCartItem1,
} from '../../test-data/shopping-cart';
import { ShoppingCartContextValue } from '../../types/shopping-cart';
import ShoppingCartContext from './shopping-cart-context';
import ShoppingCartProvider from './shopping-cart-provider';

const renderProvider = (
  component: ReactElement,
  options?: Omit<RenderOptions, 'queries'>
) => {
  const shoppingCartContext: { value?: ShoppingCartContextValue } = {};
  return {
    ...render(component, {
      wrapper: ({ children }) => {
        const Children = () => {
          shoppingCartContext.value = useContext(ShoppingCartContext);
          return children;
        };
        return (
          <ShoppingCartProvider>
            <Children />
          </ShoppingCartProvider>
        );
      },
      ...options,
    }),
    shoppingCartContext,
  };
};
describe('shopping-cart-provider', () => {
  beforeEach(() => {
    window.sessionStorage.clear();
  });

  it('should provide value', () => {
    window.sessionStorage.setItem(
      'shoppingCart',
      JSON.stringify(testShoppingCart1)
    );
    const { asFragment, shoppingCartContext } = renderProvider(
      <div>content</div>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(shoppingCartContext.value).toStrictEqual({
      shoppingCart: testShoppingCart1,
      setShoppingCart: expect.any(Function),
    });
  });

  it('should provide value when empty session storage', () => {
    const { shoppingCartContext } = renderProvider(<div>content</div>);
    expect(shoppingCartContext.value).toStrictEqual({
      shoppingCart: [],
      setShoppingCart: expect.any(Function),
    });
  });

  it('should be able to set shopping cart', () => {
    window.sessionStorage.setItem(
      'shoppingCart',
      JSON.stringify([testShoppingCartItem1])
    );
    const { shoppingCartContext } = renderProvider(<div>content</div>);
    act(() => {
      shoppingCartContext.value?.setShoppingCart(testShoppingCart1);
    });
    expect(shoppingCartContext.value).toStrictEqual({
      shoppingCart: testShoppingCart1,
      setShoppingCart: expect.any(Function),
    });
    expect(window.sessionStorage.getItem('shoppingCart')).toBe(
      JSON.stringify(testShoppingCart1)
    );
  });
});
