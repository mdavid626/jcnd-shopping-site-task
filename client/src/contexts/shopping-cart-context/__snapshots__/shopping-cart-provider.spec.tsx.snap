// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shopping-cart-provider should provide value 1`] = `
<DocumentFragment>
  <div>
    content
  </div>
</DocumentFragment>
`;
